packages/flutter/lib/src/widgets/spell_check.dart

```diff
@@ -0,0 +1,338 @@
+    String newSpanText = '';
+    bool currentSpanValid = false;
+    RegExp regex;
+
+    // Assumes that the order of spans has not been jumbled for optimization
+    // purposes, and will only search since the previously found span.
+    int searchStart = 0;
+
+    while (spanPointer < results.length) {
+      // Try finding SuggestionSpan from old results (currentSpan) in new text.
+      currentSpan = results[spanPointer];
+      currentSpanText =
+          resultsText.substring(currentSpan.range.start, currentSpan.range.end);
+
+      try {
+        // currentSpan was found and can be applied to new text.
+        newSpanText = newText.substring(
+            currentSpan.range.start + offset, currentSpan.range.end + offset);
+        currentSpanValid = true;
+      } catch (e) {
+        // currentSpan is invalid and needs to be searched for in newText.
+        currentSpanValid = false;
+      }
+
+      if (currentSpanValid && newSpanText == currentSpanText) {
+        // currentSpan was found at the same index in new text and old text
+        // (resultsText), so apply it to new text by adding it to the list of
+        // corrected results.
+        searchStart = currentSpan.range.end + offset;
+        adjustedSpan = SuggestionSpan(
+            TextRange(
+                start: currentSpan.range.start + offset, end: searchStart),
+            currentSpan.suggestions
+        );
+        correctedSpellCheckResults.add(adjustedSpan);
+      } else {
+        // Search for currentSpan in new text and if found, apply it to new text
+        // by adding it to the list of corrected results.
+        regex = RegExp('\\b$currentSpanText\\b');
+        foundIndex = newText.substring(searchStart).indexOf(regex);
+
+        if (foundIndex >= 0) {
+          foundIndex += searchStart;
+          spanLength = currentSpan.range.end - currentSpan.range.start;
+          searchStart = foundIndex + spanLength;
+          adjustedSpan = SuggestionSpan(
+              TextRange(start: foundIndex, end: searchStart),
+              currentSpan.suggestions
+          );
+          offset = foundIndex - currentSpan.range.start;
+
+          correctedSpellCheckResults.add(adjustedSpan);
+        }
+      }
+      spanPointer++;
+    }
+
+    return correctedSpellCheckResults;
+  }
+
+  /// Builds the [TextSpan] tree given the current state of the text input and
+  /// spell check results.
+  ///
+  /// The [value] is the current [TextEditingValue] requested to be rendered
+  /// by a text input widget. The [composingWithinCurrentTextRange] value
+  /// represents whether or not there is a valid composing region in the
+  /// [value]. The [style] is the [TextStyle] to render the [value]'s text with,
+  /// and the [misspelledTextStyle] is the [TextStyle] to render misspelled
+  /// words within the [value]'s text with. The [spellCheckResults] are the
+  /// results of spell checking the [value]'s text.
+  TextSpan buildTextSpanWithSpellCheckSuggestions(
```

Comment:
Could this be static too?

Reply:
If it could be static, then maybe you don't even need to instantiate this entire class, and it's more like a bunch of helper methods.